asllib/Typing.ml

```diff
@@ -1936,15 +1943,24 @@ module Annotate (C : ANNOTATE_CONFIG) = struct
     | LDI_Typed (ldi', t) ->
         let t' = annotate_type ~loc env t in
         let+ () = check_can_be_initialized_with loc env t' ty in
-        let new_env, new_ldi' = annotate_local_decl_item loc env t' ldk ldi' in
+        let new_env, new_ldi' =
+          annotate_local_decl_item loc env t' ldk ?e ldi'
+        in
         (new_env, LDI_Typed (new_ldi', t')) |: TypingRule.LDTyped
     (* End *)
     (* Begin LDVar *)
     | LDI_Var x ->
         (* Rule LCFD: A local declaration shall not declare an identifier
            which is already in scope at the point of declaration. *)
         let+ () = check_var_not_in_env loc env x in
-        let new_env = add_local x ty ldk env in
+        let env2 = add_local x ty ldk env in
+        let new_env =
+          match e with
+          | None -> env2
+          | Some e ->
+              let () = assert (ldk = LDK_Let || ldk = LDK_Constant) in
```

Comment:
I think we need to check that `e` is statically evaluable.

```
var x = 3;
let y = x;
let z = y; // z is equal to z here
x = 4;
let w = y; // w is not equal to z here
```
I guess you take care of this in `S_Decl`?
If so, isn't it better to do it here and avoid having to pass `?e` and re-check that `ldk` is `LDK_LET` or `LDK_Constant`?